src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.BaseNameProvider.cs

```diff
@@ -0,0 +1,146 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.FindSymbols
+{
+    internal partial class SymbolTreeInfo
+    {
+        /// <summary>
+        /// Used to produce the simple-full-name components of a type from metadata.
+        /// The name is 'simple' in that it does not contain things like backticks,
+        /// generic arguments, or nested type + separators.  Instead just hte name
+        /// of the type, any containing types, and the component parts of its namespace
+        /// are added.  For example, for the type "X.Y.O`1.I`2, we will produce [X, Y, O, I]
+        /// 
+        /// </summary>
+        private class BaseNameProvider : ISignatureTypeProvider<object>
+        {
+            private static char[] s_dotSeparator = { '.' };
+
+            private static readonly ObjectPool<BaseNameProvider> s_providerPool =
+                new ObjectPool<BaseNameProvider>(() => new BaseNameProvider());
+
+            private List<string> _nameParts;
+
+            private BaseNameProvider()
+            {
+            }
+
+            public static BaseNameProvider Allocate(List<string> nameParts)
+            {
+                var provider = s_providerPool.Allocate();
+                Debug.Assert(provider._nameParts == null);
+                provider._nameParts = nameParts;
+                return provider;
+            }
+
+            public static void Free(BaseNameProvider provider)
+            {
+                Debug.Assert(provider._nameParts != null);
+                provider._nameParts = null;
+                s_providerPool.Free(provider);
+            }
+
+            public object GetTypeFromDefinition(MetadataReader reader, TypeDefinitionHandle handle, byte rawTypeKind)
+            {
+                // Add the namespace and name of this type definition.  But only do this
+                // for the first type we hit.  Any further types will be things like
+                // type arguments, as we do not want to add those to the list.
+                if (_nameParts.Count == 0)
+                {
+                    var typeDefinition = reader.GetTypeDefinition(handle);
+                    var declaringType = typeDefinition.GetDeclaringType();
+                    if (declaringType.IsNil)
+                    {
+                        // Not a nested type, just add the containing namespace.
+                        AddNamespaceParts(reader, typeDefinition.NamespaceDefinition);
+                    }
+                    else
+                    {
+                        // We're a nested type, recurse and add the type we're declared in.
+                        // It will handle adding the namespace properly.
+                        GetTypeFromDefinition(reader, declaringType, rawTypeKind);
```

Comment:
>GetTypeFromDefinition(reader, declaringType, rawTypeKind); [](start = 24, length = 58)

I think it would be better to use a separate method to do the nested type recursion, rather then reuse GetTypeFromDefinition. It's confusing as it is now since GetTypeFromDefinition is usually a leaf method so the first statement ```if (_nameParts.Count==0)``` is surprising.